host/sgx/sgxquote.c

```diff
@@ -622,3 +710,236 @@ oe_result_t oe_sgx_get_supported_attester_format_ids(
+                p_qe_identity && qe_identity_size > 0 &&
+                p_qe_identity_issuer_chain && qe_identity_issuer_chain_size > 0)
+            {
+                p_sgx_endorsements = (sgx_ql_qve_collateral_t*)oe_malloc(
+                    sizeof(sgx_ql_qve_collateral_t));
+                if (p_sgx_endorsements == NULL)
+                {
+                    OE_RAISE_MSG(
+                        OE_OUT_OF_MEMORY,
+                        "Out of memory while creating SGX QVL endorsements.",
+                        NULL);
+                }
+                memset(p_sgx_endorsements, 0, sizeof(sgx_ql_qve_collateral_t));
+
+                p_sgx_endorsements->version = collateral_version;
+                p_sgx_endorsements->tcb_info = (char*)p_tcb_info;
+                p_sgx_endorsements->tcb_info_size = tcb_info_size;
+                p_sgx_endorsements->tcb_info_issuer_chain =
+                    (char*)p_tcb_info_issuer_chain;
+                p_sgx_endorsements->tcb_info_issuer_chain_size =
+                    tcb_info_issuer_chain_size;
+                p_sgx_endorsements->pck_crl = (char*)p_pck_crl;
+                p_sgx_endorsements->pck_crl_size = pck_crl_size;
+                p_sgx_endorsements->root_ca_crl = (char*)p_root_ca_crl;
+                p_sgx_endorsements->root_ca_crl_size = root_ca_crl_size;
+                p_sgx_endorsements->pck_crl_issuer_chain =
+                    (char*)p_pck_crl_issuer_chain;
+                p_sgx_endorsements->pck_crl_issuer_chain_size =
+                    pck_crl_issuer_chain_size;
+                p_sgx_endorsements->qe_identity = (char*)p_qe_identity;
+                p_sgx_endorsements->qe_identity_size = qe_identity_size;
+                p_sgx_endorsements->qe_identity_issuer_chain =
+                    (char*)p_qe_identity_issuer_chain;
+                p_sgx_endorsements->qe_identity_issuer_chain_size =
+                    qe_identity_issuer_chain_size;
+
+                OE_TRACE_INFO("SGX endorsements from OE SDK are used for quote "
+                              "verification\n");
+            }
+
+        } while (0);
+
+        err = _sgx_qv_verify_quote(
+            p_quote,
+            (uint32_t)quote_size,
+            p_sgx_endorsements,
+            expiration_check_date,
+            (uint32_t*)p_collateral_expiration_status,
+            (sgx_ql_qv_result_t*)p_quote_verification_result,
+            (sgx_ql_qe_report_info_t*)p_qve_report_info,
+            (uint32_t)supplemental_data_size,
+            p_supplemental_data);
+
+        if (p_sgx_endorsements)
+            oe_free(p_sgx_endorsements);
+
+        // To align with current quote verification logic, only accept TCB
+        // status
+        // - UpToDate
+        // - SW Hardening needed
+        result = sgx_qvl_error_to_oe(err);
+
+        if (result != OE_OK)
+        {
+            OE_RAISE_MSG(
+                OE_PLATFORM_ERROR,
```

Comment:
Is my understanding correct:
When 1.SGX_DCAP_QVL env doesn't set or QVL doesn't exist, and 2. QVL verification failed, the function returns the same error code `OE_PLATFORM_ERROR`. 
It's a safe move as even users use QVL verification and it failed, it will still perform another current OE verification so that it's guaranteed step 1 didn't break current OE verification. 
And for an invalid quote, step 1 will temporarily spend at most two verifications (QVL and OE) right? Can the error message in QVL ("SGX ECDSA Quote verification error quote3_error_t=0x{err}\n") be distinguished from OE verification error message? So user can at least tell which verification failed.

Reply:
Only when SGX_DCAP_QVL env doesn't set or QVL cannot be loaded, we will return `OE_PLATFORM_ERROR`, all other quote verification failures would use other error code, see function `sgx_qvl_error_to_oe()`.

When quote verification failed with QVL, I don't think it's a better choice to still perform quote verification with OE. In step 1, if user set the `USE_SGX_QVL`, he/she should explicitly know DCAP QVL would be used, and we should report error also.

Error message "SGX ECDSA Quote verification error quote3_error_t=0x{err}\n" is not so clear, I will update it.

Reply:
OK, I misunderstood `OE_RAISE_MSG(OE_PLATFORM_ERROR, ...`. The `result` here is actually `sgx_qvl_error_to_oe(err)`.
Thanks for clarification.

Reply:
OK, I misunderstood `OE_RAISE_MSG(OE_PLATFORM_ERROR, ....)` The `result` here is actually `sgx_qvl_error_to_oe(err)`.
Thanks for clarification.